File: mobile/lib/models/booking.dart
import 'gedung.dart';

class Booking {
  final int id;
  final String bookingCode;
  final int userId;
  final int gedungId;
  final String tanggal;
  final String? tanggalSelesai;
  final String jamMulai;
  final String jamSelesai;
  final double totalHarga;
  final String paymentType;
  final String paymentStatus;
  final String status;
  final String? paymentProof;
  final String? paymentVerifiedAt;
  final Gedung? gedung;
  final List<dynamic>? addOns;

  final String? batasBayarLunas;

  Booking({
    required this.id,
    required this.bookingCode,
    required this.userId,
    required this.gedungId,
    required this.tanggal,
    this.tanggalSelesai,
    required this.jamMulai,
    required this.jamSelesai,
    required this.totalHarga,
    required this.paymentType,
    required this.paymentStatus,
    required this.status,
    this.paymentProof,
    this.paymentVerifiedAt,
    this.batasBayarLunas,
    this.gedung,
    this.addOns,
  });

  String get statusLabel {
    switch (status) {
      case 'pending':
        return 'Menunggu';
      case 'confirmed':
        return 'Dikonfirmasi';
      case 'cancelled':
        return 'Dibatalkan';
      default:
        return status;
    }
  }

  String get paymentLabel {
    switch (paymentStatus) {
      case 'unpaid':
        return 'Belum Dibayar';
      case 'waiting':
        return 'Menunggu Verifikasi';
      case 'paid_dp':
        return 'DP 50% Lunas';
      case 'paid_lunas':
        return 'Lunas';
      default:
        return paymentStatus;
    }
  }

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'],
      bookingCode: json['booking_code'],
      userId: json['user_id'],
      gedungId: json['gedung_id'],
      tanggal: json['tanggal'],
      tanggalSelesai: json['tanggal_selesai'],
      jamMulai: json['jam_mulai'],
      jamSelesai: json['jam_selesai'],
      totalHarga: double.parse(json['total_harga'].toString()),
      paymentType: json['payment_type'],
      paymentStatus: json['payment_status'],
      status: json['status'],
      paymentProof: json['payment_proof'],
      paymentVerifiedAt: json['payment_verified_at'],
      batasBayarLunas: json['batas_bayar_lunas'],
      gedung: json['gedung'] != null ? Gedung.fromJson(json['gedung']) : null,
      addOns: json['add_ons'],
    );
  }
}

File: app/Models/Booking.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Support\Facades\Storage;

class Booking extends Model
{
    use HasFactory;

    protected function casts(): array
    {
        return [
            'tanggal'             => 'date',
            'tanggal_selesai'     => 'date',
            'jam_mulai'           => 'string',
            'jam_selesai'         => 'string',
            'payment_verified_at' => 'datetime',
            'batas_bayar_lunas'   => 'datetime',
        ];
    }

    public function isUnpaidLunas(): bool
    {
        return $this->payment_type === 'lunas' && $this->payment_status === 'unpaid' && $this->status !== 'cancelled';
    }

    public function getDpAmountAttribute(): float
    {
        return $this->total_harga * 0.5;
    }

    public function getPaymentStatusLabelAttribute(): string
    {
        return match ($this->payment_status) {
            'unpaid'     => 'Belum Dibayar',
            'waiting'    => 'Menunggu Verifikasi',
            'paid_dp'    => 'DP 50% Lunas',
            'paid_lunas' => 'Lunas',
            default      => 'Unknown',
        };
    }

    public function getPaymentProofUrlAttribute(): ?string
    {
        if (!$this->payment_proof) {
            return null;
        }
        return Storage::url($this->payment_proof);
    }

    public function isUnpaidDp(): bool
    {
        return $this->payment_type === 'dp' && $this->payment_status === 'unpaid' && $this->status !== 'cancelled';
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function gedung(): BelongsTo
    {
        return $this->belongsTo(Gedung::class);
    }

    public function addOns(): BelongsToMany
    {
        return $this->belongsToMany(AddOn::class, 'booking_add_ons')
            ->withPivot('quantity')
            ->withTimestamps();
    }
}
